### PR TITLE
feat(liquidation-sdk-viem): sky additional liquidity - fallback mechanism for sky<>mkr and dai<>usds

### PR DESCRIPTION
**Pull request by 0x666c6f**

### CODE REVIEW

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 1]
```diff
@@ -240,41 +241,159 @@ export const check = async <
...
+                        retry = false;
+                      } else {
+                        encoder
+                          .erc20Approve(srcToken, bestSwap.tx.to, srcAmount)
+                          .pushCall(
+                            bestSwap.tx.to,
+                            BigInt(bestSwap.tx.value),
+                            bestSwap.tx.data,
+                          );
+                        retry = false;
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:394]:
```suggestion
                        break;
```

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 2]
```diff
@@ -240,41 +241,159 @@ export const check = async <
...
+                          srcAmount,
+                        );
+                        encoder.erc20Approve(
+                          tries[1]?.srcToken!,
+                          bestSwap.tx.to,
+                          srcAmount,
+                        );
+
+                        encoder[conversionFunction](srcAmount, executorAddress);
+                        retry = false;
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:385]:
```suggestion
                        break;
```

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 3]
```diff
@@ -240,41 +241,159 @@ export const check = async <
...
+                              secondToken!,
+                              secondSwap.tx.to,
+                              halfAmount,
+                            )
+                            .pushCall(
+                              secondSwap.tx.to,
+                              BigInt(secondSwap.tx.value),
+                              secondSwap.tx.data,
+                            );
+                          retry = false;
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:362]:
```suggestion
                          break;
```

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 4]
```diff
@@ -240,41 +241,159 @@ export const check = async <
...
-                      .pushCall(
-                        bestSwap.tx.to,
-                        BigInt(bestSwap.tx.value),
-                        bestSwap.tx.data,
-                      );
-
+                    let retry = true;
+                    const tries: { srcAmount: bigint; srcToken: Address }[] =
+                      [];
+                    while (retry) {
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:247]:
```suggestion
                    while (true) {
```

Comment (Rubilmax) [reply packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:247]: With `break` statements

Ideally also you could export the whole `while` loop to a function so it is more readable (and maybe you can take advantage of early returns in this case

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 5]
```diff
@@ -240,41 +241,159 @@ export const check = async <
...
+                        ) {
+                          srcToken = Sky.getAlternativeToken(srcToken);
+                          retry = true;
+                        }
+                        // If even using the alternative token we still don't have enough liquidity, we try with both tokens (and half the amount)
+                        else if (
+                          Sky.isTokenPair(
+                            tries[0]?.srcToken,
+                            tries[1]?.srcToken,
+                          )
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:292]: If there are 2 tries (implied at this stage), it is necessarily a Sky token pair isn't it?

Comment (Rubilmax) [reply packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:292]: In which case this simplifies to an `else` statement and the `isTokenPair` util is not useful anymore (we can get rid of it)

**Review by Rubilmax — DISMISSED** [review]

File: packages/liquidation-sdk-viem/src/tokens/sky.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import { mainnetAddresses } from "@morpho-org/liquidation-sdk-viem";
+import type { Address } from "viem";
+
+export namespace Sky {
+  export const MKR_SKY_CONVERTER = "0xBDcFCA946b6CDd965f99a839e4435Bcdc1bc470B";
+  export const DAI_USDS_CONVERTER =
+    "0x3225737a9Bbb6473CB4a45b7244ACa2BeFdB276A";
```
[the remark this answers is not in the document]
Comment (0x666c6f) [reply packages/liquidation-sdk-viem/src/tokens/sky.ts:7]: Yeah it's fine no worries, tbh I hesitated between both. I'll change it

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts (hunk 5 above)
Comment (0x666c6f) [reply packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:292]: yes but the previous `if` can also fail if the token is not a Sky related token

Comment (Rubilmax) [reply packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:292]: Alright so we just need to check `tries.length == 2` right?

Comment (0x666c6f) [reply packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:292]: Technically yes, as of now this is the only situation where we have 2 tries. I just thought it was "cleaner" to check it as well, but I don't mind removing it if you prefer

File: packages/liquidation-sdk-viem/src/LiquidationEncoder.ts [hunk 7]
```diff
@@ -1,4 +1,8 @@
 import { type Address, MathLib } from "@morpho-org/blue-sdk";
+import {
+  daiUsdsConverterAbi,
+  mkrSkyConverterAbi,
+} from "@morpho-org/liquidation-sdk-viem/src/abis.js";
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/src/LiquidationEncoder.ts]:
```suggestion
} from "./abis.js";
```

File: packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts [hunk 8]
```diff
@@ -174,7 +174,9 @@ export const check = async <
                   collateralUnderlyingAsset ?? market.params.collateralToken;
                 let srcAmount = withdrawnAssets ?? seizedAssets;
 
-                const encoder = new LiquidationEncoder(executorAddress, client);
+                const encoder: LiquidationEncoder<
+                  Client<Transport, Chain, Account>
+                > = new LiquidationEncoder(executorAddress, client);
```
Comment (Rubilmax) [inline packages/liquidation-sdk-viem/examples/whitelisted-erc4626-1inch.ts:179]:
```suggestion
                const encoder = new LiquidationEncoder(executorAddress, client);
```

**Review by Rubilmax — APPROVED** [review]

**Pull request merged**